File: src/db/test_data/content.js
const casual = require('casual');
const times = require('../utils/times');
const PyShell = require('../utils/python_shell');
const content_types = ['content' /*, 'debate', 'squad', 'task'*/];

const get_texts_from_corpus = (
    limit = 30,
    min_text_length = 10,
    max_text_length = 250
) => {
    return new Promise((resolve, reject) => {
        PyShell.runPython(PyShell.PYTHON_METHODS[1], [
            JSON.stringify({
                limit: limit,
                min_text_length: min_text_length,
                max_text_length: max_text_length,
            }),
        ])
            .then(result => {
                resolve({
                    bodies_cache: result.bodies,
                    titles_cache: result.titles,
                });
                return;
            })
            .catch(err => {
                console.log(err);
                throw new Error(err);
                reject(err);
            });
    });
};

const contents_vectorize = contents => {
    return new Promise((resolve, reject) => {
        PyShell.runPython(PyShell.PYTHON_METHODS[3], [
            JSON.stringify({
                contents: contents,
            }),
        ])
            .then(results => {
                resolve({
                    contents: results.map((result, index) => {
                        let val = contents[index];
                        val.vector = result.vector;
                        return val;
                    }),
                });
                return;
            })
            .catch(err => {
                console.log(err);
                throw new Error(err);
                reject(err);
            });
    });
};

const content = (
    key,
    users_limit,
    isCorpus = true,
    bodies_cache = [],
    titles_cache = []
) => {
    const mode = content_types[key % content_types.length];
    const score = casual.double((from = 0), (to = 1000));
    return {
        user_id: casual.integer((from = 1), (to = users_limit)),
        parent_id: null,
        score: score,
        pure_score: score,
        last_payout_score: score,
        last_payout_pure_score: score,
        author_score: score * 0.9,
        author_pure_score: score * 0.9,
        last_payout_author_score: score * 0.9,
        last_payout_author_pure_score: score * 0.9,
        voters_score: score * 0.1,
        voters_pure_score: score * 0.1,
        last_payout_voters_score: score * 0.1,
        last_payout_voters_pure_score: score * 0.1,
        vector: JSON.stringify(
            Array.apply(null, Array(50)).map(function() {
                return 0;
            })
        ),
        url: '',
        title: isCorpus ? titles_cache[key] : casual.title,
        body: isCorpus
            ? bodies_cache[key]
            : casual.sentences((n = Math.floor(Math.random() * 20))),
        locale: 'ja',
        country_code: 'JP',
        meta: '',
        path: `/${key + 1}`,
        count: 2,
        isStory: true,
        isNsfw: false,
        ishide: false,
        allowEdit: true,
        allowDelete: true,
        allowReply: true,
        allow_votes: true,
        allow_curation_rewards: true,
        max_accepted_payout: '',
        hasPendingPayout: false,
        deadline_cashout_at: new Date(),
        last_score_at: new Date(),
        last_payout_at: new Date(),
        isOpinionWanted: key % 6 == 0,
        isRequestWanted: false,
        isPrivate: false,
        valid: true,
        permission: true,
        created_at: new Date(),
        updated_at: new Date(),
    };
};

async function contents(limit = 30, users_number = 30, isCorpus = true) {
    let contents_array = [];
    const result = isCorpus
        ? await get_texts_from_corpus(limit).catch(err => {
              throw new Error(err);
          })
        : {};
    await times(limit)(() => {
        contents_array.push(
            content(
                contents_array.length,
                users_number,
                isCorpus,
                result.bodies_cache || [],
                result.titles_cache || []
            )
        );
    });

    // contents_array = await contents_vectorize(contents_array).catch((e) => {
    //     throw new Error(e)
    // })

    return contents_array;
}

module.exports = contents;

File: src/python-scripts/models/content.py
from models.upvote import Upvotes, Upvote
import re
from collections import OrderedDict
import sys, json
import numpy as np

class Contents(object):
    def __init__(self, JSON):
        self.items = [Content(content) for content in JSON if content is not None]

    def toOwnedBaseDict(self):
        return { content.id : [ upvote.VoterId for upvote in content.upvotes.items ] for content in self.items if len(content.upvotes.items) > 0 }

    # def toSortArrayWithComment(self):
    #     return { val.rootId : [ item.UserId for item in self.getsByRootId(val.rootId) ] for val in self.items }

    def toJSON(self):
        return { "contents": [content.toJSON() for content in self.items] }

    def getById(self, id):
        for item in self.items:
            if item.id == id:
                return item
            continue

    def getsByUserId(self, id):
        return [item for item in self.items if item.UserId == id]

    # def getsByRootId(self, rootId):
    #     return [item for item in self.items if item.rootId == rootId]


class Content(object):
    p = re.compile(r"<[^>]*?>")
    def __init__(self, JSON):
        if JSON is None:
            return
        self.id = JSON.get('id')
        self.UserId = JSON.get('UserId')
        self.ParentId = JSON.get('ParentId')
        self.RequestId = JSON.get('RequestId')
        self.score = JSON.get('score')
        self.pure_score = JSON.get("pure_score")
        self.author_score = JSON.get('author_score')
        self.author_pure_score = JSON.get("author_pure_score")
        self.voters_score = JSON.get('voters_score')
        self.voters_pure_score = JSON.get("voters_pure_score")
        self.url = JSON.get('url')
        self.title = JSON.get('title')
        self.body = JSON.get('body')
        self.meta = JSON.get('meta')
        self.mode = JSON.get('mode')
        self.vector =  json.loads(JSON.get("vector")) if isinstance(JSON.get("vector"), str) else JSON.get("vector")
        self.isStory = JSON.get('isStory')
        self.isNsfw = JSON.get('isNsfw')
        self.ishide = JSON.get('ishide')
        self.allowEdit = JSON.get('allowEdit')
        self.allowDelete = JSON.get('allowDelete')
        self.allowReply = JSON.get('allowReply')
        self.allow_votes = JSON.get('allow_votes')
        self.allow_curation_rewards = JSON.get('allow_curation_rewards')
        self.max_opinions_select = JSON.get('max_opinions_select')
        self.pending_payout_value = JSON.get('pending_payout_value')
        self.curator_payout_value = JSON.get('curator_payout_value')
        self.max_accepted_payout = JSON.get('max_accepted_payout')
        self.total_payout_value = JSON.get('total_payout_value')
        self.total_pending_payout_value = JSON.get('total_pending_payout_value')
        self.hasPendingPayout = JSON.get('hasPendingPayout')
        self.cashout_time = JSON.get('cashout_time')
        self.deadline_cashout_time = JSON.get('deadline_cashout_time')
        self.last_payout_at = JSON.get('last_payout_at')
        self.isPrivate = JSON.get('isPrivate')
        self.valid = JSON.get('valid')
        self.permission = JSON.get('permission')
        self.upvotes = Upvotes(JSON.get("UpVotes")) if "UpVotes" in JSON else Upvotes()

    @property
    def pure_body(self):
        return self.remove_html_tag(self.body)

    @property
    def pure_array_body(self):
        return self.remove_html_tag(self.body).split(r'\n||。||\.')

    # @property
    # def rootId(self):
    #     m = re.search(r"\/\d+$", self.path if self.path not None else '')
    #     return m.group(0) if m else self.id


    def remove_html_tag(self, html_text):
        return self.p.sub("", html_text)

    def toJSON(self):
        return {
            'id': self.id,
            'UserId': self.UserId,
            'ParentId': self.ParentId,
            'score': self.score,
            'pure_score': self.pure_score,
            'url': self.url,
            'title': self.title,
            'body': self.body,
            'meta': self.meta,
            'mode': self.mode,
            'vector': self.vector if True not in [np.isnan(vec) for vec in self.vector] else np.zeros(len(self.vector)),
            'isStory': self.isStory,
            'isNsfw': self.isNsfw,
            'ishide': self.ishide,
            'allowEdit': self.allowEdit,
            'allowDelete': self.allowDelete,
            'allowReply': self.allowReply,
            'allow_votes': self.allow_votes,
            'allow_curation_rewards': self.allow_curation_rewards,
            'max_opinions_select': self.max_opinions_select,
            'pending_payout_value': self.pending_payout_value,
            'curator_payout_value': self.curator_payout_value,
            'max_accepted_payout': self.max_accepted_payout,
            'total_payout_value': self.total_payout_value,
            'total_pending_payout_value': self.total_pending_payout_value,
            'hasPendingPayout': self.hasPendingPayout,
            'cashout_time': self.cashout_time,
            'deadline_cashout_time': self.deadline_cashout_time,
            'last_payout_at': self.last_payout_at,
            'isPrivate': self.isPrivate,
            'valid': self.valid,
            'permission': self.permission,
        }
